README.md
# Genetic-Algoritm

This implementation of Genetic Algorithm is based on the ideas from the article:
[Genetic Algorithm on Polynomials](https://www.researchgate.net/publication/269132907_Optimal_Polynomial_Regression_Models_by_using_a_Genetic_Algorithm)

### Usage:

```
stack install
Genetic-Algoritm-exe <Num of GEnerations> <Size of population> <Threshold fitness> <elite size>
```
Currently the code tries to find Rosenbrock function from the above Article.

Notice that if the Algirithm reaches the threshold before the final genretation it will return the output and stop. For example this is a test run from my PC (the +RTS flag uses to work on 8 cores):

```
» Genetic-Algoritm-exe 500 20 0.01 10 +RTS -N8                                                                                             orenm@OrenPC

No. of generatoins: 500
Population Size: 20
Score of goal: 0.01
Elite size (N-elit): 10

Progress:
[===================================================================================================================>.............................................................................................]
 200.012 * x1^2 * x2 + 100.488 * x1^4 + (-2.669) * x1 + 99.998 * x2^2 + 0.891 * x1^(-1) * x2^(-3)
Fitness of final Polynomial is: 5.796397023532152e-3

Attention: component with coefficient smaller tan 0.001 are rounded to nothing
------------------------------------------------------------
» 
```

#### Contact me
If you have any questions or problems, or if you want to take this project a step forward

orenm13@gmail.com


ChangeLog.md
# Changelog for Genetic-Algoritm

## Unreleased changes


src/Genetic.hs
{-# LANGUAGE FlexibleContexts     #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeSynonymInstances #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE ScopedTypeVariables  #-}

module Genetic where

import           Numeric.LinearAlgebra
import           System.Random.Shuffle    (shuffleM)
import           System.Random
import           System.ProgressBar
import           Control.Monad
import           Control.Monad.IO.Class
import           Control.Concurrent.Async (mapConcurrently)
import qualified Data.Vector as V
import           Data.List
import           Data.Maybe

class (Ord a, Eq a, Show a) => Individual a where
    fitness :: a -> IO (Maybe Double)
    mutateIndividual :: a -> IO a
    crossover :: (a, a) -> IO a

runGeneticAlgorithm :: Individual a
                    => Double      -- Goal
                    -> Int         -- maxNumIterations
                    -> Int         -- eliteSize
                    -> [a]         -- Initial (random) genetation
                    -> ProgressRef -- Progress Bar
                    -> IO a
runGeneticAlgorithm threshold
                    maxNumIterations
                    eliteSize
                    population
                    progRef   = do
    finalGroup <- foldM (makeGeneration threshold eliteSize progRef)
                        initialGeneration
                        [1..maxNumIterations]
    best finalGroup
    where
        initialGeneration = population

makeGeneration :: Individual a
               => Double      -- Goal
               -> Int         -- eliteSize
               -> ProgressRef
               -> [a]         -- population
               -> Int         -- generation number
               -> IO [a]
makeGeneration goal eliteSize progRef population _ = do
    cleanPop <- filterM (fmap isJust . fitness) population
    mFitness <- fitness =<< best cleanPop
    let bestFitness = fromMaybe (error "Invalid Fitness") mFitness
    if bestFitness < goal
    then return cleanPop
    else do incProgress progRef 1
            nextGeneration cleanPop eliteSize


nextGeneration :: Individual a
               => [a]    -- population
               -> Int    -- elite size
               -> IO [a]
nextGeneration population eliteSize = do
    children    <- makeChildren population eliteSize
    elita <- elite population eliteSize
    return (elita ++ children)

best :: Individual a
     => [a]
     -> IO a
best pop = do
    fitnesses <- mapConcurrently fitness pop
    let combined = filter (isJust . fst) $ zip fitnesses pop
    --             [(fitness, individual)]
        chosen   = minimumBy (\x y -> compare (fst x) (fst y)) combined
    return $ snd chosen

mutate :: Individual a => [a] -> IO [a]
mutate = mapConcurrently mutateIndividual

elite :: Individual a
      => [a]
      -> Int
      -> IO [a]
elite individuals nElit = do
    fitnesses <- mapConcurrently fitness individuals
    let combined   = zip fitnesses individuals
        results    = map snd . take nElit . sortOn fst $ combined
        mFitnesses = map fst . take nElit . sortOn fst $ combined
    return results

selectCouples :: Individual a
           => [a] -- population
           -> Int -- number of couples
           -> IO [(a, a)]
selectCouples population numOfCouples =
    sequence . take numOfCouples $ repeat $ selectCouple population


selectCouple :: Individual a
             => [a]
             -> IO (a, a)
selectCouple population = do
    fitnesses <- mapConcurrently fitness population
    let combined    = filter (isJust . fst) $ zip fitnesses population
        population_ = map snd . sortOn (fromJust . fst) $ combined
        indexList   = [0..(length population_ - 1)]
        duplicated  = duplicateByValue indexList
    [index1, index2] <- take 2
                      . nub
                    <$> shuffleM duplicated
    return ( population_ !! index1
           , population_ !! index2
           )

duplicateByValue :: [Int]
                 -> [Int]
duplicateByValue = foldr (\ x -> (++) (replicate x x)) []

makeChildren :: Individual a
             => [a]    -- population
             -> Int    -- nElit
             -> IO [a]
makeChildren population nElit = do
    parents   <- selectCouples population $ length population - nElit
    children  <- mapConcurrently crossover parents
    fitnesses <- mapConcurrently fitness children
    mutate $ map fst $ filter (isJust . snd) $ zip children fitnesses

bestFit :: Individual a
        => [a]
        -> IO Double
bestFit pop = minimum . catMaybes <$> mapConcurrently fitness pop


src/Formulas.hs
module Formulas where

import           Numeric.LinearAlgebra
import           Control.Monad.IO.Class
import           Control.Monad (replicateM)
import           System.Random
import qualified Data.Vector as V
import           Data.List

type X          = Matrix R
type Beta       = Vector R
type Y          = Vector R
type Block      = V.Vector Int
type Powers     = V.Vector Block
type Input      = V.Vector Double
type Inplist    = V.Vector Input
type Polynomial = (Beta, X)


-- V.Vectors, Vectors and Matrices Transformators
vvToM :: V.Vector (V.Vector Double)
      -> Matrix Double
vvToM = fromLists . V.toList . V.map V.toList

mToVV :: Matrix Double
      -> V.Vector (V.Vector Double)
mToVV = V.fromList . map V.fromList . toLists
-----------------------------------------------------

-- (5) in the article --
f :: Powers
  -> Input
  -> V.Vector Double
f pss inp = V.map oneBlock pss
    where
        oneBlock ps = V.product $ V.zipWith (**) inp $ V.map fromIntegral ps

calcX :: Inplist
      -> Powers -> X
calcX inputs pss = vvToM (V.map (f pss) inputs)

-- (~7) in the article --
cost :: Beta
     -> X
     -> Y
     -> Double
cost b x y = abs $ sumElements $ e**2
    where e = y - x #> b

-- (8) in the article --
normEq :: X
       -> Y
       -> Beta
normEq x y  = (inv (tr x <> x) <> tr x) #> y :: Beta
---------------- Show polynomial ---------------------
showPolynomial :: Beta
               -> Powers
               -> String
showPolynomial coeffs powers =
    intercalate " +" $ zipWith showCoeff coeffs' powers'
        where
            coeffs' = filter isNotZero $ toList coeffs
            powers' = V.toList $ V.map V.toList powers
            showCoeff coeff powerList = " " ++ show_ (round3 coeff) ++ matchVarsPowers powerList
            isNotZero x = round3 x /= 0

matchVarsPowers :: [Int] -> String
-- Example: [1, 4, 2] => x1 * x2^4 * x3^2
matchVarsPowers powers =
    intercalate "" $ zipWith varToPower varNums powers
        where
            varNums = [1..length powers]
            varToPower :: Int -> Int -> String
            varToPower ind 0   = ""
            varToPower ind 1   = " * x" ++ show ind
            varToPower ind pow = " * x" ++ show ind ++ "^" ++ show_ pow

-- show_ :: Num n => n -> n
show_ num = if num > 0 then show num else "(" ++ show num ++ ")"

    -------------------------------------------------------
randomPowerGen :: Int      -- m
               -> Int      -- n
               -> IO (V.Vector (V.Vector Int)) -- m >< n Random Powers
randomPowerGen m n = do
    lists <- replicateM m $ replicateM n $ randomRIO (-4, 4)
    return $ V.fromList $ map V.fromList lists

round3 :: Double -> Double
round3 x = (fromInteger $ round $ x * (10^3)) / (10.0^^3)


src/Polynomials.hs
{-# LANGUAGE FlexibleInstances   #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Polynomials where

import           Numeric.LinearAlgebra
import           System.Random hiding (rand)
import           Control.Monad.IO.Class
import           Control.Monad
import           Control.Exception
import           Data.Maybe
import qualified Data.Vector as V
import qualified Data.Array.Accelerate as A
import           Genetic
import           Formulas

import Debug.Trace

data Article = Article
            { _articlePowers  :: Powers
            , _articleInputs  :: Inplist
            , _articleOutputs :: Y
            } deriving (Eq, Ord)

instance Show Article where
      show = show . _articlePowers

instance Individual Article where
   -- fitness :: a -> IO (Maybe Double)
      fitness (Article ps is os) = do
            let x' = calcX is ps
                rows = length $ toLists x'
                cols = length . head $ toLists x'
                y = os
            noise_ <- rand rows cols
            let   noise = noise_ * 100.0
                  x = if det (tr x' <> x') == 0
                      then x' + noise
                      else x'
            -- bug fix for BLAS matrix rank on big numbers
            return $ if det (tr x <> x) == 0
                     then Nothing
                     else Just $ cost (normEq x y) x y

   -- mutateIndividual :: a -> IO a
      mutateIndividual (Article pows _a _b) = do
            r <- randomIO
            if mutationProbability < r
            then inner pows
            else return $ Article pows _a _b
            where
                  mutationProbability :: Double
                  mutationProbability = 0.1

                  inner :: Powers -> IO Article
                  inner x = do
                        blockIndex <- randomRIO (0, V.length x - 1)
                        powerIndex <- randomRIO (0, V.length (x V.! 0) - 1)
                        newPower   <- randomRIO (-4, 4)

                        let updated = updateVV x ( blockIndex, powerIndex) newPower
                        return $ Article updated _a _b


   -- crossover :: (a, a) -> IO a
      crossover ( Article pows1 _a _b
                , Article pows2 _ _
                ) = do
            index <- randomRIO (0, V.length pows1)
            let leftPart  = V.take index pows1
            let rightPart = V.drop index pows2
            return $ Article (V.concat [leftPart, rightPart]) _a _b


updateVV :: V.Vector (V.Vector a) -- what to update
         -> (Int, Int)            -- where
         -> a                     -- what to put there
         -> V.Vector (V.Vector a) -- result
updateVV vvs (xIdx, yIdx) value =
      vvs V.// [(xIdx, updatedV)]
      where
            updatedV  = originalV V.// [(yIdx, value)]
            originalV = vvs V.! xIdx

app/Main.hs
module Main where

import           Numeric.LinearAlgebra
import           System.Environment (getArgs)
import           System.Random
import           System.ProgressBar
import           Control.Monad (replicateM)
import           Control.Monad.IO.Class
import           Data.Maybe
import           Data.List
import qualified Data.Vector as V
import           Polynomials
import           Genetic
import           Formulas

main :: IO ()
main = do
    (nGenerations:nPop:goal:nElit:_) <- getArgs

    ------------ Information ------------
    putStrLn ""
    putStrLn $ "No. of generatoins: "  ++ nGenerations
    putStrLn $ "Population Size: "     ++ nPop
    putStrLn $ "Score of goal: "       ++ goal
    putStrLn $ "Elite size (N-elit): " ++ nElit
    putStrLn ""

    ------------ Progress Bar -----------
    putStrLn "Progress:"
    (pb, asy) <- startProgress (msg "") (msg "") 10 (Progress 0 (read nGenerations))

    ------------ Algorithm --------------
    powers <- randomPowerGen 6 2
    let
        ------ Inps and outps of function ------
        inps  = inpsRosen
        outps = fromList $ V.toList yRosen
        ----------------------------------------
        individual = Article powers inps outps
        init = replicate (read nPop) individual

    algo <- runGeneticAlgorithm (read goal) (read nGenerations) (read nElit) init pb
    let
        finalPowers = _articlePowers algo
        x = calcX inps finalPowers
        y = outps
        coeffs = normEq x y
        ----------------------------------------
    grade <- show . fromJust <$> fitness algo

    putStrLn ""
    putStrLn $ showPolynomial coeffs finalPowers
    putStrLn $ "Fitness of final Polynomial is: " ++ grade
    putStrLn ""
    putStrLn "Attention: component with coefficient smaller tan 0.001 are rounded to nothing"


openVLists = V.map V.toList

-- Check x + 1:
simple x = x + 1.0

inpsSimple :: V.Vector Double
inpsSimple = V.fromList
       [ 1
       , 2
       , 3
       , 4]

powsSimple :: V.Vector Double
powsSimple = V.fromList
            [1.0,
             0]

ySimple = V.map simple inpsSimple

-- Check x**2 + y**2 -1:
circle [x, y] = x**2 + y**2 --  -1

inpsCircle :: Inplist
inpsCircle = V.fromList
     . map V.fromList
     $ [ [0.0001, 1.1]
       , [2, 0.001]
       , [2, 1.05]
       , [1, 2.11]
       , [0.1, 2]]

powsCircle :: Powers
powsCircle = V.fromList
          . map V.fromList
          $ [[2, 0],
             [0, 2],
             [0, 0]]

yCircle = V.map circle $ openVLists inpsCircle

-- Check Rosenbrock:
rosen [x, y] =
    1 - 2*x + x**2 + 100*y**2 + 200*x**2*y + 100*x**4

inpsRosen :: Inplist
inpsRosen = V.fromList
     . map V.fromList
     $ [[i, j] | i <- [-0.9, -0.8 .. -0.2], j <- [0.2, 0.3 .. 0.9]]


powsRosen :: Powers
powsRosen = V.fromList
        . map V.fromList
        $ [[0, 0],
           [1, 0],
           [2, 0],
           [0, 2],
           [2, 1],
           [4, 0]]

yRosen = V.map rosen $ openVLists inpsRosen
